File: contracts/test/script/compare-storage-layout-deployed.js
const { execSync } = require("child_process");

// Different types of types have different IDs, so we need to strip the ID from the type
// (e.g., t_contract(LightClient)44013 → t_contract(LightClient))
function normalizeType(type) {
  const end = type.indexOf(")");

  if (end !== -1) {
    return type.slice(0, end + 1);
  }

  return type;
}

// Extracts the storage layout from a deployed contract using cast
function extractDeployedLayout(contractAddress) {
  try {
    // Use cast with --json flag to get structured storage layout
    const result = execSync(
      `cast storage ${contractAddress} --rpc-url ${process.env.RPC_URL || "http://localhost:8545"} --json`,
      {
        encoding: "utf8",
        stdio: ["pipe", "pipe", "pipe"],
      },
    );

    const output = result.toString().trim();

    // Handle empty output
    if (!output) {
      return [];
    }

    const data = JSON.parse(output);

    // Handle empty storage array
    if (!data.storage || data.storage.length === 0) {
      return [];
    }

    // Convert to our format
    return data.storage.map(({ label, slot, offset, type }) => ({
      label,
      slot: parseInt(slot),
      offset: parseInt(offset),
      type: normalizeType(type),
    }));
  } catch (error) {
    // Check if stderr contains "Storage layout is empty" warning
    if (error.stderr && error.stderr.toString().includes("Storage layout is empty")) {
      return [];
    }
    console.error(`Error getting deployed contract layout: ${error.message}`);
    throw error;
  }
}

// Extracts the storage layout using `forge inspect` and parses the JSON output
function extractLocalLayout(contractName) {
  const output = execSync(`forge inspect ${contractName} storageLayout --json`).toString();

  // Find the JSON part by looking for the first '{' and last '}'
  const startIndex = output.indexOf("{");
  const lastIndex = output.lastIndexOf("}");

  if (startIndex === -1 || lastIndex === -1) {
    throw new Error(`No valid JSON found in output: ${output}`);
  }

  const jsonOutput = output.substring(startIndex, lastIndex + 1);
  const layout = JSON.parse(jsonOutput);
  return layout.storage.map(({ label, slot, offset, type }) => ({
    label,
    slot: parseInt(slot),
    offset: parseInt(offset),
    type: normalizeType(type),
  }));
}

// Compare two storage layout arrays
// expects the first layout to be the deployed one and the second to be the new one
function compareLayouts(layoutA, layoutB) {
  const errors = [];

  if (layoutA.length > layoutB.length) {
    // the new layout should have same or more variables
    errors.push(
      `Deployed contract has ${layoutA.length} storage variables but new contract has only ${layoutB.length}`,
    );
    return { compatible: false, errors };
  }

  for (let i = 0; i < layoutA.length; i++) {
    const a = layoutA[i];
    const b = layoutB[i];

    if (a.label !== b.label) {
      errors.push(`Position ${i}: label mismatch - deployed: "${a.label}", local: "${b.label}"`);
    }
    if (a.slot !== b.slot) {
      errors.push(`Position ${i} (${a.label}): slot mismatch - deployed: ${a.slot}, local: ${b.slot}`);
    }
    if (a.offset !== b.offset) {
      errors.push(`Position ${i} (${a.label}): offset mismatch - deployed: ${a.offset}, local: ${b.offset}`);
    }
    // Compare types, but allow contract/interface substitution
    if (a.type !== b.type) {
      // Extract base type for contracts and interfaces
      const aIsContract = a.type.startsWith("t_contract(");
      const bIsContract = b.type.startsWith("t_contract(");

      // If both are contracts/interfaces, they're compatible (both store as address)
      if (!(aIsContract && bIsContract)) {
        errors.push(`Position ${i} (${a.label}): type mismatch - deployed: ${a.type}, local: ${b.type}`);
      }
    }
  }

  return { compatible: errors.length === 0, errors };
}

const [deployedAddress, newContractName] = process.argv.slice(2);

if (!deployedAddress || !newContractName) {
  console.error("Usage: node compare-storage-layout-deployed.js deployedAddress newContractName");
  console.error("Example: node compare-storage-layout-deployed.js 0x123...abc LightClientV3");
  process.exit(1);
}

try {
  const deployedLayout = extractDeployedLayout(deployedAddress);
  const newLayout = extractLocalLayout(newContractName);

  // If deployed layout is empty (e.g., OZ V5 namespaced storage), skip checks
  if (deployedLayout.length === 0) {
    console.log(true);
    process.exit(0);
  }

  const { compatible, errors } = compareLayouts(deployedLayout, newLayout);

  if (!compatible) {
    console.error(`\nStorage layout mismatch between deployed contract (${deployedAddress}) and ${newContractName}:`);
    console.error("\nErrors:");
    errors.forEach((err) => console.error(`  - ${err}`));

    console.error("\nDeployed layout:");
    deployedLayout.forEach((item, i) => {
      console.error(`  [${i}] ${item.label}: slot=${item.slot}, offset=${item.offset}, type=${item.type}`);
    });

    console.error(`\nLocal layout (${newContractName}):`);
    newLayout.forEach((item, i) => {
      console.error(`  [${i}] ${item.label}: slot=${item.slot}, offset=${item.offset}, type=${item.type}`);
    });
    console.error("");
  }

  console.log(compatible);
  process.exit(0);
} catch (err) {
  console.error("Error comparing layouts:", err.message);
  process.exit(1);
}

File: contracts/test/script/compare-storage-layout-deployed.py
#!/usr/bin/env python3
import argparse
import json
import os
import subprocess
import sys


def normalize_type(t: str) -> str:
    idx = t.find(")")
    return t[: idx + 1] if idx != -1 else t


def to_int(v) -> int:
    if isinstance(v, int):
        return v
    s = str(v)
    return int(s, 16) if s.startswith(("0x", "0X")) else int(s)


def extract_deployed_layout(address: str) -> list | None:
    rpc_url = os.environ["RPC_URL"]
    result = subprocess.run(
        ["cast", "storage", address, "--rpc-url", rpc_url, "--json"],
        capture_output=True,
        text=True,
    )
    if "Storage layout is empty" in result.stderr or "Storage layout is empty" in result.stdout:
        return None  # signal: skip check
    if result.returncode != 0:
        print(result.stderr, file=sys.stderr)
        sys.exit(1)
    raw = result.stdout.strip()
    if not raw or raw == "null":
        return None
    data = json.loads(raw)
    storage = data.get("storage") or []
    if not storage:
        return None
    return [
        {
            "label": e["label"],
            "slot": to_int(e["slot"]),
            "offset": to_int(e["offset"]),
            "type": normalize_type(e["type"]),
        }
        for e in storage
    ]


def extract_local_layout(contract: str) -> list:
    try:
        result = subprocess.run(
            ["forge", "inspect", contract, "storageLayout", "--json"],
            capture_output=True,
            text=True,
            check=True,
        )
    except subprocess.CalledProcessError as e:
        print(e.stderr, file=sys.stderr)
        sys.exit(1)
    data = json.loads(result.stdout)
    return [
        {
            "label": e["label"],
            "slot": to_int(e["slot"]),
            "offset": to_int(e["offset"]),
            "type": normalize_type(e["type"]),
        }
        for e in (data.get("storage") or [])
    ]


def types_compatible(t_a: str, t_b: str) -> bool:
    if t_a == t_b:
        return True
    # Both contract/interface types store as address — compatible
    return t_a.startswith("t_contract(") and t_b.startswith("t_contract(")


def compare(a: list, b: list) -> bool:
    # New contract may add storage slots at the end — that's safe for upgrades.
    if len(a) > len(b):
        return False
    return all(
        a[i]["label"] == b[i]["label"]
        and a[i]["slot"] == b[i]["slot"]
        and a[i]["offset"] == b[i]["offset"]
        and types_compatible(a[i]["type"], b[i]["type"])
        for i in range(len(a))
    )


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="Compare deployed contract storage layout against local source. RPC_URL must be set in environment."
    )
    parser.add_argument("deployed_address", help="Deployed contract address (0x...)")
    parser.add_argument("local_contract", help="Local contract name (forge inspect format)")
    args = parser.parse_args()

    deployed = extract_deployed_layout(args.deployed_address)
    if deployed is None:
        print("true", end="")
        sys.exit(0)
    local = extract_local_layout(args.local_contract)
    print("true" if compare(deployed, local) else "false", end="")
